org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -176,6 +177,15 @@
      */
     protected static HashMap<PK11Cert, SSLFDProxy> serverTemplates = new HashMap<PK11Cert, SSLFDProxy>();
 
+    /**
+     * Whether or not the session cache has been initialized already.
+     *
+     * A session cache must always be created in order to utilize a
+     * server-side JSSEngine. However, NSS isn't threadsafe when creating
+     * such a cache, so synchronize it within JSSEngine.
+     */
+    private final static AtomicBoolean sessionCacheInitialized = new AtomicBoolean();
```

Comment:
The `AtomicBoolean` should work fine, but a regular `boolean sessionCacheInitialized` and a `synchronized initializeSessionCache()` probably would work as well. It's up to you, whichever simpler.